crates_io_cdn_logs/src/fastly/mod.rs

```diff
@@ -0,0 +1,170 @@
+mod json;
+
+use crate::paths::parse_path;
+use crate::DownloadsMap;
+use std::borrow::Cow;
+use tokio::io::{AsyncBufRead, AsyncBufReadExt};
+use tracing::{debug_span, instrument, warn};
+
+#[instrument(level = "debug", skip(reader))]
+pub async fn count_downloads(reader: impl AsyncBufRead + Unpin) -> anyhow::Result<DownloadsMap> {
+    let mut downloads = DownloadsMap::new();
+
+    let mut lines = reader.lines();
+    while let Some(line) = lines.next_line().await? {
+        let span = debug_span!("process_line");
+        let _guard = span.enter();
+
+        let Some(json) = parse_line(&line) else {
+            warn!("Failed to find JSON start");
+            continue;
+        };
+
+        let json = match parse_json(json) {
+            Ok(json) => json,
+            Err(error) => {
+                warn!("Failed to parse JSON: {error}");
+                continue;
+            }
+        };
+
+        if json.method() != "GET" {
+            // Ignore non-GET requests.
+            continue;
+        }
+
+        if json.status() != 200 {
+            // Ignore non-200 responses.
+            continue;
+        }
+
+        let url = decode_url(json.url());
+
+        // We're avoiding parsing to `url::Url` here for performance reasons.
+        // Since we're already filtering out non-200 responses, we can assume
+        // that the URL is valid.
+
+        let Some((name, version)) = parse_path(&url) else {
+            continue;
+        };
+
+        let date = json.date_time().date_naive();
+
+        downloads.add(name, version, date);
+    }
+
+    Ok(downloads)
+}
+
+#[instrument(level = "debug", skip(line))]
+fn parse_line(line: &str) -> Option<&str> {
+    // A regex could also be used here, but the `find()` call appears to
+    // be roughly 10x faster.
+    line.find(r#"]: {"#).map(|pos| &line[pos + 3..])
```

Comment:
I had a little bit of trouble finding the log line format we use, which is https://docs.fastly.com/en/guides/changing-log-line-formats#classic-format — could we include that in a comment wherever you feel is appropriate?

Reply:
> I had a little bit of trouble finding the log line format we use

you and me both, which is why I didn't include any links 😂